torch_uncertainty/losses.py

```diff
@@ -173,3 +173,198 @@ def forward(
+        loss_type: Optional[str] = "log",
+        reduction: Optional[str] = "mean",
+    ) -> None:
+        super().__init__()
+
+        if num_classes < 0:
+            raise ValueError(
+                "The number of classes should be non-negative, but got "
+                f"{num_classes}."
+            )
+        self.num_classes = num_classes
+
+        if reg_weight and (reg_weight < 0):
+            raise ValueError(
+                "The regularization weight should be non-negative, but got "
+                f"{reg_weight}."
+            )
+        self.reg_weight = reg_weight
+
+        if annealing_step and (annealing_step <= 0):
+            raise ValueError(
+                "The annealing step should be positive, but got "
+                f"{annealing_step}."
+            )
+        self.annealing_step = annealing_step
+
+        if reduction != "none" and reduction != "mean" and reduction != "sum":
+            raise ValueError(f"{reduction} is not a valid value for reduction.")
+        self.reduction = reduction
+
+        if loss_type != "mse" and loss_type != "log" and loss_type != "digamma":
+            raise ValueError(
+                f"{loss_type} is not a valid value for mse/log/digamma loss."
+            )
+        self.loss_type = loss_type
+
+    def _mse_loss(self, evidence: Tensor, targets: Tensor) -> Tensor:
+        evidence = torch.relu(evidence)
+        alpha = evidence + 1.0
+        strength = torch.sum(alpha, dim=1, keepdim=True)
+
+        loglikelihood_err = torch.sum(
+            (targets - (alpha / strength)) ** 2, dim=1, keepdim=True
+        )
+        loglikelihood_var = torch.sum(
+            alpha * (strength - alpha) / (strength * strength * (strength + 1)),
+            dim=1,
+            keepdim=True,
+        )
+        loss = loglikelihood_err + loglikelihood_var
+
+        return loss
+
+    def _log_loss(self, evidence: Tensor, targets: Tensor) -> Tensor:
+        evidence = torch.relu(evidence)
+        alpha = evidence + 1.0
+        strength = alpha.sum(dim=-1, keepdim=True)
+
+        loss = torch.sum(
+            targets * (torch.log(strength) - torch.log(alpha)),
+            dim=1,
+            keepdim=True,
+        )
+
+        return loss
+
+    def _digamma_loss(self, evidence: Tensor, targets: Tensor) -> Tensor:
+        evidence = torch.relu(evidence)
+        alpha = evidence + 1.0
+        strength = alpha.sum(dim=-1, keepdim=True)
+
+        loss = torch.sum(
+            targets * (torch.digamma(strength) - torch.digamma(alpha)),
+            dim=1,
+            keepdim=True,
+        )
+
+        return loss
+
+    def _kldiv_reg(
+        self,
+        evidence: Tensor,
+        targets: Tensor,
+    ) -> Tensor:
+        evidence = torch.relu(evidence)
+        alpha = evidence + 1.0
+
+        kl_alpha = (alpha - 1) * (1 - targets) + 1
+
+        ones = torch.ones(
+            [1, self.num_classes], dtype=torch.float32, device=evidence.device
```

Comment:
the dtype and number of classes should be automatically determined

Reply:
Yes, you are right! Thanks

Reply:
Since it is also used in the one hot function, I think leave it in the class is better

Reply:
Yes, of course, I just meant that the users will have to use partial(DECLoss, num_classes=...) which is something that I would have preferred to avoid...